### ISSUE TITLE
NetworkManager plugin?

### ISSUE THREAD

**Issue by dark-penguin** [ancwrd1/snx-rs#58]
Hi!
Thank you very much for this project; I thought I was going to be SOL using Linux at my new job, but I found this, and it "just works"! :heart: 

I have a pretty bold suggestion/request - but who knows, maybe it would turn out possible: it would be really cool to have this as a NetworkManager plugin.

Why?
- There would be no need to maintain a GUI since NetworkManager already has every capability you can think of, and then some.
- Configuration would be "familiar" for those who is already using other supported VPN types. Having only one familiar "network" icon in the notification area is a minor plus.
- Seems like it can do a lot of things on its own, like assigning addresses, injecting DNS resolvers, etc.
- Most importantly for me, NetworkManager now has this cool capability to inject DNS resolvers and zones into dnsmasq, in order to avoid touching resolv.conf . This way, I can have multiple per-zone resolvers, instead of querying VPN-provided resolvers first, getting NXDOMAIN from them for my other local zones, and therefore being unable to access the rest of my network.

I don't know if it's possible to write a plugin in Rust though. Could be a "fun" project, or could be unpleasant/impossible.

Here's one example I've found; a plugin for IPSec (StrongSWAN) that seems to be just a wrapper - so a similar plugin could replace the GUI but still use the Rust implementation under the hood.
https://download.strongswan.org/NetworkManager/

Oh, but I don't know if it supports popup windows for MFA challenges... I've been previously using OpenVPN with MFA, but they had a "hacky" setup where your MFA code had to be appended to your password, and it was only checked once in a while, so it was not a big problem. Here, your MFA is going to be checked every time, so such a solution wouldn't work. But I was able to reach NetworkManager developers (for an unrelated issue), so maybe they could add the necessary capability on their part, or suggest how to do it if it's already there.

**Comment by ancwrd1** [comment]
That would be nice to have as an additional module but I am not familiar with NM architecture, and it requires a significant time investment. May be someone else could look into it in the near future.

**Comment by ancwrd1** [comment]
This looks great! Would it work for SAML SSO authentication via web browser (identity_provider authentication). It requires to launch a system browser to a given URL and await for the OTP from the browser GET request.
Additionally, I would also consider adding support for NM DNS resolver, in addition to the existing systemd-resolved and the resolv.conf.

**Comment by ancwrd1** [comment]
> I'll investigate this further. Which approach would you prefer?

Thanks for a great summary. I would prefer system browser. The embedded WebView has several issues in addition to the webkit dependency:
* It is considered less secure for OAuth workflows (https://developers.googleblog.com/upcoming-security-changes-to-googles-oauth-20-authorization-endpoint-in-embedded-webviews/) and might be blocked by some IdPs.
* It requires a separate UI to manage user data (cookies), e.g. to force the user to log out.
* It does not integrate nicely with global SSO state which the user may already have, for example with EntraID or Google.

The current implementation uses webkit only when a `mobile-access` feature flag is enabled and when the "Mobile Access" authentication is used, for some legacy configurations. It is necessary because we need to extract the OTP from the web page when it finishes loading, using Javascript.

> This one is relatively straightforward, if im not mistaken code and the infrastructure is already in place.

I think there are two use cases for NM-based DNS management:
* The NM plugin
* Standalone application (the current implementation)

The most important part here is the split-DNS, which is a must have feature for the corporate tunnel. Also I am not sure if NM actually supports routing domains, may be something worth investigating.

For the NM plugin use case, we can indeed utilize the proposed `no_dns=true` and don't let snxcore configure DNS resolver. For the standalone app, I would probably look if it is possible to integrate it to NetworkManager if it is detected as a DNS resolver (/etc/resolv.conf pointing to /run/NetworkManager/resolv.conf stub). Currently it will consider it as a plain resolv.conf and will modify the file which isn't a clean approach. I am actually not sure if it is possible to let NM manage the DNS in this case, if the network interface is not managed by NM. I think I tried it at some point without much success.

> We can make this the default behavior or add a configurable parameter like nm-dns = native | snxcore if backward compatibility is required.

That would be good, in case if NM still leaks private DNS requests to the tunnel. `systemd-resolved` never does that if routing domains are used.

**Comment by diegodario88** [comment]
Hi @ancwrd1,

Here's an update on both fronts:

---

##  DNS Management 

The NM plugin already implements the approach you suggested:

**Current implementation** (`apps/nm-snx-service/src/config.rs:118-123`):
```rust
params.no_dns = true;              // snxcore doesn't touch DNS
params.set_routing_domains = true; // Enable ~ prefix for split DNS
```

**Ip4Config signal** (`apps/nm-snx-service/src/main.rs:791-848`):
- DNS servers passed via `dns` key
- Search domains prefixed with `~` for routing domains (split DNS)
- `dns-priority = -100` for proper VPN DNS preference

This allows NetworkManager to handle DNS natively through its systemd-resolved/dnsmasq integration.

**Regarding standalone + NM as resolver**: I agree this is a separate investigation. Currently, if `/etc/resolv.conf` points to NM's stub, snxcore treats it as a plain resolv.conf. Integrating via NM's D-Bus API for non-managed interfaces would require further research.

---

## SAML/SSO via System Browser

**What already exists in snxcore:**
- `SystemBrowser` (`crates/snxcore/src/browser.rs`) - uses `opener::open()` to launch system browser
- `OtpListener` (`crates/snxcore/src/otp.rs`) - HTTP server on `127.0.0.1:7779` with 120s timeout
- `MfaType::IdentityProvider` detection in `tunnel/ssl/connector.rs`

**What needs to be added to nm-snx-service:**

1. Add `opener` dependency to `nm-snx-service/Cargo.toml`
2. In `do_connect()`, detect `MfaType::IdentityProvider` in `PendingChallenge`
3. When detected:
   - Start `OtpListener` 
   - Open system browser with the IdP URL via `opener::open()`
   - Await OTP from the listener (120s timeout)
   - Continue with `connector.challenge_code(otp)`
4. No auth-dialog needed for this flow (browser handles everything)

**Proposed flow:**
```
User clicks Connect → NM calls ConnectInteractive
  → Authenticate returns PendingChallenge(IdentityProvider, "https://idp.example.com/...")
  → nm-snx-service detects MfaType::IdentityProvider
  → Starts OtpListener on 127.0.0.1:7779
  → Opens system browser with IdP URL
  → User authenticates in browser
  → IdP redirects to http://127.0.0.1:7779/{otp}
  → OtpListener captures OTP
  → connector.challenge_code(otp)
  → Connection completes
```

---

##  Testing Challenge

I don't have access to a VPN server with SAML/Identity Provider authentication to test this implementation. 

Do you have any suggestions for testing? Possibilities I can think of:
1. A test/mock IdP setup that simulates the flow
2. Unit tests with mocked OtpListener responses
3. If you have access to such a server, perhaps you could test a PR?

Let me know your thoughts on the implementation plan and testing approach!

**Comment by ancwrd1** [comment]
@diegodario88 I can test IdP login, yes. The OtpListener itself is already unit-tested, I think there is no need to add another one.
I have some plans to add proper E2E tests with the emulated VPN server (https://github.com/ancwrd1/cp-ikev1-proxy) and integrate into the CI/CD, but didn't have time yet to implement them and the cp-ikev1-proxy is not yet finalized.

**Comment by nexcode** [comment]
Guys, if you finish this, you will be my best friends!

**Comment by Felipe-53** [comment]
Love this project! How can we sponsor you? Asking nothing back, as you guys have already done so much for us!

**Comment by cojack** [comment]
@diegodario88 I built the same thing independently before finding this issue — a NM VPN service
plugin plus, on top of it, the **KDE/plasma-nm** side. Your branch is clearly further along and
better structured, so rather than compete I would like to offer what does not overlap, plus two
concrete findings from testing against a live gateway.

## The KDE half is missing, and it is not optional for Plasma users

plasma-nm gates Connect on the presence of **its own** UI plugin, before NetworkManager is ever
consulted. With the `.name` descriptor installed and the service provably loadable, the Plasma
applet still refuses with "Plasma is missing support for 'snx-rs' VPN connections." So
`nm-snx-editor` for GNOME Settings does nothing for a KDE user — a `VpnUiPlugin` subclass is
required, and that is a separate C++ artifact.

What I have working against a live gateway:

* a `VpnUiPlugin` subclass — configuration widget plus authentication widget
* the login type populated by querying the gateway, so it is a select rather than free text
* MFA prompting through `SecretsRequired`, with the gateway's own wording shown in the dialog
* NetworkManager applying the address and split DNS itself, verified with `resolvectl`

Happy to contribute it against your branch if you want the coverage.

## Finding 1: prompt string matching breaks on non-English gateways

```rust
if prompt_lower.contains("password") && !params.password.is_empty() {
    info!("Auto-submitting password");
```

Check Point sends **every** text factor as `MfaType::PasswordInput`; only its position
distinguishes the stored password from a one-time code. My gateway's prompts are localised — the
second factor arrives as `+Wpisz kod FortiToken` — so any gateway that also localises the
*password* prompt will not match here, the stored password will never be auto-submitted, and the
user gets asked for factor 1 while believing they are being asked for a token. Whatever they type
then goes to the gateway as the password.

`snxcore`'s own controller solves this by counting rather than reading
(`controller.rs`, `mfa_index` against `params.password_factor`):

```rust
if !params.password.expose_secret().is_empty() && self.mfa_index == params.password_factor {
    Ok(params.password.expose_secret().to_owned())
}
```

Mirroring that makes factor selection locale-independent. I hit this exact bug and it cost me
several hours before I understood the prompt was for factor 1.

## Finding 2: `never-default` is hardcoded true

```rust
// Tell NM not to add default route - snxcore handles routing
ip4_config.insert("never-default", Value::new(true));
```

That is right for split tunnelling but wrong for a full-route configuration, where the gateway
does want the default route. `ConnectionInfo` already carries the flag, so
`Value::new(!info.default_route)` covers both.

## Two things your implementation does better than mine

* **Event-driven emission.** You emit `Config`/`Ip4Config`/`STARTED` from a spawned task fed by
  `TunnelEvent::Connected`. I polled `GetStatus` instead, which is both uglier and hit a real trap:
  emitting those signals from *inside* the `ConnectInteractive` or `NewSecrets` method call leaves
  NetworkManager rewriting its own state when the call returns, then waiting forever for a
  `STARTED` it has already seen. Your design avoids that by construction.
* **`dns-priority: -100`.** I had not set it.

## One architectural question worth deciding deliberately

Your service links `snxcore` and runs the tunnel in-process. Mine drives the existing
command-mode daemon over its IPC socket instead. The trade-off: with the in-process design,
`snxctl status` and `snx-rs-gui` cannot see a tunnel that NetworkManager started, so a user has
two independent notions of "connected". Driving the daemon keeps one owner and all frontends
agree, at the cost of requiring the service to be running.

Not arguing for either — but it determines whether NM becomes an alternative frontend or a
parallel one, and it is easier to settle now than later.

## Also worth knowing: NetworkManager cannot own routing

I tested full variant B — NM owning addresses, routes and DNS. Addresses and DNS work.
Routes cannot, because `ConnectionInfo` exposes no routes for the plugin to hand over, so the
tunnel ends up with only its on-link prefix. On my gateway the pushed DNS servers sit outside that
prefix, so nothing resolved and the keepalive tore the tunnel down within twenty seconds. Letting
the daemon keep routing — which your comment already does — is the only option unless the IPC is
extended.
